lib/elixir/src/elixir_quote.erl

```diff
@@ -254,6 +256,19 @@ is_valid(context, Context) -> is_atom(Context) andalso (Context /= nil);
 is_valid(generated, Generated) -> is_boolean(Generated);
 is_valid(unquote, Unquote) -> is_boolean(Unquote).
 
+shallow_validate_ast(Expr) ->
+  case shallow_valid_ast(Expr) of
+    true -> Expr;
+    false -> argument_error(<<"tried to unquote invalid AST: ", ('Elixir.Kernel':inspect(Expr))/binary>>)
+  end.
+
+shallow_valid_ast(Expr) when is_list(Expr) -> lists:all(fun shallow_valid_ast/1, Expr);
+shallow_valid_ast(Expr) when is_atom(Expr); is_binary(Expr); is_number(Expr); is_pid(Expr) -> true;
+shallow_valid_ast({Left, Right}) -> shallow_valid_ast(Left) andalso shallow_valid_ast(Right);
```

Comment:
I am thinking we should skip this one... the combination of traversing lists and two-element tuples mean we can traverse large keywords? 🤔 Or perhaps we pass a flag to control the level, so we don't recur beyond the first 2-elem tuple/list?

Reply:
Keywords was what I had in mind, I didn't think about large ones (is it so common in AST?)
I like the idea of having some nested flag to avoid doing a deep check 👍 will update

Reply:
Keywords was what I had in mind, I didn't think about large ones (is it so common in AST?)
I like the idea of having some nested flag to avoid doing a deep check 👍 will update

Reply:
Fixed in [96bb33a](https://github.com/elixir-lang/elixir/pull/13950/commits/96bb33a7349906941f2e5f8a1bdc52f9c50446b5).
- we can now only check one depth of lists
- I don't think having deeply nested size-2 tuples is such a realistic scenario, so I'm not checking for the depth there to keep it simple